Installation issue: dealii forces -std=c++17 on all users

### Steps to reproduce the issue

```console
spack spec -I dealii%gcc@11
[+]  dealii@9.6.2+adol-c+arborx+arpack+assimp+cgal~complex~cuda~doc+examples+examples_compile+ginkgo+gmsh+gsl+hdf5~int64~ipo+kokkos+metis+mpi+muparser~nanoflann~netcdf~oce+opencascade~optflags+p4est+petsc+platform-introspection~python+scalapack+simplex+slepc+sundials+symengine+taskflow+threads+trilinos+vtk build_system=cmake build_type=DebugRelease cxxstd=17 generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^adol-c@2.7.2~advanced_branching+atrig_erf~boost+doc+examples~openmp~sparse+stdczero+traceless_refcounting build_system=autotools patches:=715b878 arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]      ^arborx@1.3~cuda~ipo+mpi~openmp~rocm+serial~sycl+trilinos build_system=cmake build_type=Release cxxstd=20 generator=make arch=linux-ubuntu22.04-cascadelake %cxx=gcc@11.4.0
[+]      ^arpack-ng@3.9.1~icb~ipo+mpi+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^assimp@6.0.2~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]          ^cmake@3.25.2~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-cascadelake 
[e]          ^pkgconf@1.6.3 build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[+]      ^boost@1.88.0+atomic~charconv+chrono~clanglibcpp~cobalt+container~context~contract~coroutine+date_time~debug+exception~fiber+filesystem+graph~graph_parallel+icu+iostreams~json+locale+log+math~mpi~mqtt5+multithreaded~nowide~numpy~pic+program_options~python+random+regex+serialization+shared~signals2~singlethreaded~stacktrace+system~taggedlayout+test+thread+timer~type_erasure~url~versionedlayout+wave build_system=generic cxxstd=17 patches:=a440f96 visibility=hidden arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]          ^bzip2@1.0.8~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]          ^icu4c@76.1 build_system=autotools cxxstd=17 arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]              ^python@3.10.12+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches:=0d98e93,7d40923,ebdca64,f2fd060 arch=linux-ubuntu22.04-cascadelake 
[+]          ^xz@5.6.3~pic build_system=autotools libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^zstd@1.5.7+programs build_system=makefile compression:=none libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]      ^cgal@5.6~core~demos+eigen~imageio~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^eigen@3.4.0~ipo~nightly~rocm build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^gmp@6.3.0+cxx build_system=autotools libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^mpfr@4.2.1 build_system=autotools libs:=shared,static patches:=3ec29a6 arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^autoconf-archive@2023.02.20 build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[+]              ^texinfo@7.2~xs build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                  ^gettext@0.23.1+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                      ^tar@1.35 build_system=autotools zip=pigz arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                          ^pigz@2.8 build_system=makefile arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[e]      ^cmake@3.31.8~doc+ncurses+ownlibs~qtgui build_system=generic build_type=Release arch=linux-ubuntu22.04-cascadelake 
[+]      ^compiler-wrapper@1.0 build_system=generic arch=linux-ubuntu22.04-cascadelake 
[e]      ^gcc@11.4.0~binutils+bootstrap~graphite~nvptx~piclibs~profiled~strip build_system=autotools build_type=RelWithDebInfo languages:='c,c++,fortran' arch=linux-ubuntu22.04-cascadelake 
[+]      ^gcc-runtime@11.4.0 build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]      ^ginkgo@1.9.0~cuda~develtools~full_optimizations+half_precision~hwloc~ipo~mpi+openmp~rocm~sde+shared~sycl build_system=cmake build_type=Release commit=20cfd68795f58078898da9890baa311b46845a8b generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]      ^glibc@2.35 build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[e]      ^gmake@4.3~guile build_system=generic patches:=599f134 arch=linux-ubuntu22.04-cascadelake 
[+]      ^gmsh@4.13.1+alglib~cairo+cgns+compression~eigen~external+fltk+gmp~hdf5~ipo+med+metis+mmg~mpi+netgen~oce+opencascade~openmp~petsc~privateapi+shared~slepc~tetgen+voropp build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]          ^cgns@4.5.0~base_scope~fortran+hdf5~int64~ipo~legacy~mem_debug+mpi~pic+scoping+shared~static~testing~tools build_system=cmake build_type=Release generator=make patches:=0ecd9e4 arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^fltk@1.3.7+gl+shared~xft build_system=generic arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]              ^libx11@1.8.12 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^libxcb@1.17.0 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^libpthread-stubs@0.5 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^libxau@1.0.12 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^libxdmcp@1.1.5 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^xcb-proto@1.17.0 build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[+]                  ^xtrans@1.6.0 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^freetype@2.13.2+pic+shared build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^glx@1.4 build_system=bundle arch=linux-ubuntu22.04-cascadelake 
[+]              ^mesa@25.0.5+glx+llvm+opengl~opengles+osmesa~strip build_system=meson buildtype=release default_library:=shared arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                  ^bison@3.8.2~color build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                  ^flex@2.6.4+lex~nls build_system=autotools patches:=f8b85a0 arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                      ^help2man@1.49.3 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^glproto@1.4.17 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^libunwind@1.8.1~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs~pic+tests+weak_backtrace~xz~zlib build_system=autotools components:=none libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                  ^libxshmfence@1.3.3 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^llvm@20.1.8+clang~cuda~flang+gold~ipo+libomptarget~libomptarget_debug~link_llvm_dylib+lld+lldb+llvm_dylib+lua~mlir+offload+polly~python~split_dwarf~z3~zstd build_system=cmake build_type=Release compiler-rt=runtime generator=ninja libcxx=runtime libunwind=runtime openmp=runtime shlib_symbol_version=none targets:=all version_suffix=none arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]                      ^binutils@2.38~debuginfod~gas+gold+headers~interwork+ld~libiberty~lto~nls~pgo+plugins build_system=autotools compress_debug_sections=zlib libs:=shared,static arch=linux-ubuntu22.04-cascadelake 
[+]                      ^libffi@3.4.8 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                      ^lua@5.3.6+shared build_system=makefile fetcher=curl arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                          ^unzip@6.0 build_system=makefile patches:=881d2ed,f6f6236 arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                      ^perl-data-dumper@2.173 build_system=perl arch=linux-ubuntu22.04-cascadelake 
[+]                      ^swig@4.1.1 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                          ^pcre2@10.44~jit+multibyte+pic build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^meson@1.8.2 build_system=python_pip patches:=0f0b1bd arch=linux-ubuntu22.04-cascadelake 
[+]                      ^py-pip@25.1.1 build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]                      ^py-setuptools@80.9.0 build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]                      ^py-wheel@0.45.1 build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]                      ^python-venv@1.0 build_system=generic arch=linux-ubuntu22.04-cascadelake 
[e]                  ^ninja@1.10.0+re2c build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]                  ^py-mako@1.2.4 build_system=python_pip arch=linux-ubuntu22.04-cascadelake 
[+]                      ^py-markupsafe@2.1.3 build_system=python_pip arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^py-pyyaml@6.0.2+libyaml build_system=python_pip arch=linux-ubuntu22.04-cascadelake 
[+]                      ^libyaml@0.2.5 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^py-cython@3.0.12 build_system=python_pip arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                  ^xrandr@1.5.3 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^libxrandr@1.5.4 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                          ^renderproto@0.11.1 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^libxrender@0.9.12 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^randrproto@1.5.0 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^libjpeg-turbo@3.0.4~ipo~jpeg8~partial_decoder+pic build_system=cmake build_type=Release generator=make libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]              ^nasm@2.16.03 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^libpng@1.6.47~ipo~pic build_system=cmake build_type=Release generator=make libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^med@5.0.0+api23~doc~fortran~ipo+mpi~python~shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]          ^mesa-glu@9.0.2 build_system=autotools patches:=3d03e55 arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^mmg@5.8.0~doc~elas~ipo~private_headers+scotch+shared~vtk build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]              ^scotch@7.0.7+compression~esmumps~int64~ipo~metis+mpi~mpi_thread~noarch+shared+threads build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^gsl@2.8~external-cblas+pic+shared build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]      ^hdf5@1.12.3~cxx+fortran+hl~ipo~java+mpi+shared~szip~threadsafe+tools api=default build_system=cmake build_type=Release generator=make patches:=f42732a arch=linux-ubuntu22.04-cascadelake %c,fortran=gcc@11.4.0
[+]      ^intel-tbb@2022.0.0~ipo+shared+tm build_system=cmake build_type=Release cxxstd=default generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^hwloc@2.11.1~cairo~cuda~gl~level_zero~libudev+libxml2~nvml~opencl+pci~rocm build_system=autotools libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]              ^libpciaccess@0.17 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[e]              ^ncurses@6.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[+]      ^metis@5.1.0~gdb~int64~ipo~no_warning~real64+shared build_system=cmake build_type=Release generator=make patches:=4991da9,93a7903,b1225da arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]          ^gcc@11.4.0~binutils+bootstrap~graphite~nvptx~piclibs~profiled~strip build_system=autotools build_type=RelWithDebInfo languages:='c,c++,fortran' arch=linux-ubuntu22.04-cascadelake 
[+]          ^gcc-runtime@11.4.0 build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]      ^muparser@2.3.4~ipo~openmp+samples+shared~wide_char build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]      ^netlib-scalapack@2.2.2~ipo~pic+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,fortran=gcc@11.4.0
[+]      ^openblas@0.3.30~bignuma~consistent_fpcsr+dynamic_dispatch+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=none arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^opencascade@7.8.1+application_framework+data_exchange+draw~ffmpeg~freeimage~freetype~ipo+modeling_algorithms+modeling_data~rapidjson~tbb~tk+visualization~vtk build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^libxext@1.3.6 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[e]              ^util-macros@1.19.3 build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[+]              ^xextproto@7.3.0 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^xproto@7.0.31 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^libxi@1.7.10 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^fixesproto@5.0 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^inputproto@2.3.2 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^libxfixes@5.0.3 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^libxmu@1.2.1 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^libxt@1.3.1 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^kbproto@1.0.7 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^libice@1.1.2 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^libsm@1.2.6 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^util-linux-uuid@2.41 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^tcl@8.6.12 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]      ^openmpi@5.0.8+atomics~cuda~debug+fortran~gpfs~internal-hwloc~internal-libevent~internal-pmix~ipv6~java~lustre~memchecker~openshmem~rocm~romio+rsh~static~two_level_namespace+vt+wrapper-rpath build_system=autotools fabrics:=none romio-filesystem:=none schedulers:=none arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[e]          ^autoconf@2.69 build_system=autotools patches:=7793209 arch=linux-ubuntu22.04-cascadelake 
[e]          ^automake@1.16.1 build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[+]          ^libevent@2.1.12+openssl build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[e]              ^openssl@1.1.1f~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-cascadelake 
[e]          ^libtool@2.4.6 build_system=autotools arch=linux-ubuntu22.04-cascadelake 
[+]          ^numactl@2.0.18 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^openssh@9.9p1+gssapi build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]              ^krb5@1.21.3+shared build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                  ^findutils@4.10.0 build_system=autotools patches:=440b954 arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                      ^gettext@0.23.1+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                          ^libxml2@2.13.5~http+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                              ^zlib-ng@2.2.4+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                          ^tar@1.35 build_system=autotools zip=pigz arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                              ^pigz@2.8 build_system=makefile arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^libedit@3.1-20240808 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^libxcrypt@4.4.38~obsolete_api build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[e]          ^perl@5.30.0+cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu22.04-cascadelake 
[+]          ^pmix@5.0.5~munge~python build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]      ^p4est@2.8.7+mpi~openmp build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]      ^petsc@3.23.4~X~batch~cgns~complex~cuda~debug+double+examples~exodusii~fftw+fortran~giflib+hdf5~hpddm~hwloc+hypre~int64~jpeg~knl~kokkos~libpng~libyaml~memkind+metis~mkl-pardiso~mmg~moab~mpfr+mpi~mumps~openmp~p4est~parmmg~ptscotch~random123~rocm~saws~scalapack+shared~strumpack~suite-sparse+superlu-dist~sycl~tetgen~trilinos~valgrind~zoltan build_system=generic clanguage=C memalign=none arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]          ^diffutils@3.10 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^libiconv@1.18 build_system=autotools libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^hypre@2.33.0~caliper~complex~cuda~debug+fortran~gptune~gpu-aware-mpi~gpu-profiling~int64~internal-superlu+lapack~magma~mixedint+mpi~openmp~rocm+shared~superlu-dist~sycl~umpire~unified-memory build_system=autotools precision=double arch=linux-ubuntu22.04-cascadelake %c,fortran=gcc@11.4.0
[+]          ^parmetis@4.0.3~gdb~int64~ipo+shared build_system=cmake build_type=Release generator=make patches:=4f89253,50ed208,704b84f arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]          ^python@3.10.12+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches:=0d98e93,7d40923,ebdca64,f2fd060 arch=linux-ubuntu22.04-cascadelake 
[+]          ^superlu-dist@9.1.0~cuda~int64~ipo~openmp+parmetis~rocm+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^slepc@3.23.2+arpack~blopex~cuda~hpddm~rocm build_system=generic arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^suite-sparse@7.8.3~cuda~graphblas~openmp+pic build_system=generic arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[e]          ^m4@1.4.18+sigsegv build_system=autotools patches:=3877ab5,fc9b616 arch=linux-ubuntu22.04-cascadelake 
[+]      ^sundials@6.7.0+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL~cuda+examples+examples-install~f2003~fcmix+generic-math~ginkgo~hypre~int64~ipo~klu~kokkos~kokkos-kernels~lapack~magma~monitoring+mpi~openmp~petsc~profiling~pthread~raja~rocm+shared+static~superlu-dist~superlu-mt~sycl~trilinos build_system=cmake build_type=Release cstd=99 cxxstd=14 generator=make logging-level=2 logging-mpi=OFF precision=double arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^symengine@0.9.0~boostmp~flint~ipo~llvm+mpc+mpfr~openmp~piranha+shared+thread_safe build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^cereal@1.3.2~ipo build_system=cmake build_type=Release generator=make patches:=2dfa0bf arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^mpc@1.3.1 build_system=autotools libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]      ^taskflow@3.7.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^trilinos@13.4.1~adelus~adios2+amesos+amesos2+anasazi+aztec~basker+belos~boost~chaco~complex~cuda~cuda_constexpr~cuda_rdc~debug~dtk+epetra+epetraext~epetraextbtf~epetraextexperimental~epetraextgraphreorderings~exodus+explicit_template_instantiation~float+fortran~gtest~hdf5~hypre+ifpack+ifpack2~intrepid~intrepid2~ipo~isorropia+kokkos~mesquite~minitensor+ml+mpi+muelu~mumps~nox~openmp~pamgen~panzer~phalanx~piro~python~rocm~rocm_rdc~rol~rythmos+sacado~scorec~shards+shared~shylu~stk~stokhos~stratimikos~strumpack~suite-sparse~superlu~superlu-dist~teko~tempus~test~thyra+tpetra~trilinoscouplings~wrapper~x11~zoltan~zoltan2 build_system=cmake build_type=Release cxxstd=14 generator=make gotype=long_long arch=linux-ubuntu22.04-cascadelake %c,cxx,fortran=gcc@11.4.0
[+]      ^vtk@9.0.3~examples~ffmpeg~ipo+mpi+opengl2~python~qt~xdmf build_system=cmake build_type=Release generator=make patches:=0546696,2d38712,87656ca,ff9dce4 arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^double-conversion@3.3.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^expat@2.7.1+libbsd build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]              ^libbsd@0.12.2 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]                  ^libmd@1.1.0 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^gl2ps@1.4.2~doc~ipo+png+zlib build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^glew@2.2.0~ipo build_system=cmake build_type=Release generator=make patches:=7992e52,e65236d arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^jsoncpp@1.9.6~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %cxx=gcc@11.4.0
[+]          ^libogg@1.3.6 build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^libtheora@1.2.0~doc~examples build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^libtiff@4.7.0+ccitt~ipo~jbig+jpeg~jpeg12~lerc~libdeflate+logluv~lzma+lzw+next~old-jpeg~opengl+packbits~pic~pixarlog+shared+thunder~webp+zlib~zstd build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^libxml2@2.13.5~http+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^lz4@1.10.0+pic build_system=makefile libs:=shared,static arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^netcdf-c@4.9.2+blosc~byterange~dap~fsync~hdf4~jna~logging+mpi~nczarr_zip+optimize~parallel-netcdf+pic+shared+szip+zstd build_system=autotools patches:=0161eb8 arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^c-blosc@1.21.6+avx2~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]                  ^snappy@1.2.1~ipo+pic+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]              ^libaec@1.1.4~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]          ^proj@7.2.1+curl~ipo~pic+shared+tiff build_system=cmake build_type=Release generator=make patches:=cef7302,dc620ff arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[e]              ^curl@7.81~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs:=shared,static tls:=openssl arch=linux-ubuntu22.04-cascadelake 
[+]          ^pugixml@1.14~ipo+pic+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %cxx=gcc@11.4.0
[+]          ^sqlite@3.46.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]              ^readline@8.3 build_system=autotools patches:=21f0a03 arch=linux-ubuntu22.04-cascadelake %c=gcc@11.4.0
[+]          ^utf8cpp@4.0.6~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
[+]      ^zlib-ng@2.2.4+compat~new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-cascadelake %c,cxx=gcc@11.4.0
```


### Error message

There is no error message from spack, the package installs succesfully. (I'm not sure I categorized this report correctly...)

The installed cmake targets file at lib/cmake/deal.II/deal.IITargets.cmake sets -std=c++17. This forces every user of the package to use c++17. It's OK that deal.ii is built with c++17, but I would like to use c++20 in my project. If deal.II was only using the compile features to set the c++ standard (in the same targets.cmake file `INTERFACE_COMPILE_FEATURES "cxx_std_17"`), then it would be possible for users of the installed package to use a higher c++ standard.

This is an example project that uses deal.ii:
```cmake
cmake_minimum_required(VERSION 3.25)
project(testd2 LANGUAGES CXX)

find_package(deal.II CONFIG REQUIRED)

add_executable(testd2 main.cpp)
target_link_libraries(testd2 PRIVATE dealii::dealii)
target_compile_options(testd2 PRIVATE -std=c++20)
```
```c++
#include <numbers>

int main()
{
    return std::numbers::pi_v<double>;
}
```

compile error:
```
testd2/main.cpp:5:12: error: use of undeclared identifier 'std'
    5 |     return std::numbers::pi_v<double>;
      |            ^
testd2/main.cpp:5:37: error: expected '(' for function-style cast or type construction
    5 |     return std::numbers::pi_v<double>;
      |                               ~~~~~~^
testd2/main.cpp:5:38: error: expected expression
    5 |     return std::numbers::pi_v<double>;
      |                                      ^
3 errors generated.
gmake[2]: *** [CMakeFiles/testd2.dir/build.make:79: CMakeFiles/testd2.dir/main.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:87: CMakeFiles/testd2.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
```


### Information on your system

* **Spack:** 1.0.1 (https://github.com/spack/spack/commit/17d34285729c1bb39dfbdef09c5b78d88bb9e3f9)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/32c54f0906004d7fd1f72fd1b5970bf2bf094e26
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-cascadelake

### Additional information

I have not found a way to override the flag (using normal CMake, there's probably a way to hack it by manually parsing the list of flags). The deal.II flag always ends up behind my c++20 flag in the command line. The easiest workaround I found is to edit the targets.cmake file manually and remove the flag.

My reason to believe that it is the deal.II spack package that sets the flag is that a manual installation of deal.II (outside spack but using the same dependencies) does not set the flag. I wasn't able to test it since I don't know how to use an edited package.py. But I suspect it's these lines in the dealii package.py: 
```
        # Enforce the specified C++ standard
        if spec.variants["cxxstd"].value != "default":
            cxxstd = spec.variants["cxxstd"].value
            cxx_flags.extend(["-std=c++{0}".format(cxxstd)])
```

@jppelteret  @luca-heltai

Note: No build log since the build of deal.ii was successful.

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

--- a/repos/spack_repo/builtin/packages/dealii/package.py
+++ b/repos/spack_repo/builtin/packages/dealii/package.py
@@ -499,7 +499,7 @@ class Dealii(CMakePackage, CudaPackage):
         # Enforce the specified C++ standard
         if spec.variants["cxxstd"].value != "default":
             cxxstd = spec.variants["cxxstd"].value
-            cxx_flags.extend(["-std=c++{0}".format(cxxstd)])
+            options.append(self.define("CMAKE_CXX_STANDARD", "{0}".format(cxxstd)))
 
         # Performance
         # Set recommended flags for maximum (matrix-free) performance, see

--- a/repos/spack_repo/builtin/packages/dealii/package.py
+++ b/repos/spack_repo/builtin/packages/dealii/package.py
@@ -498,8 +498,7 @@ class Dealii(CMakePackage, CudaPackage):
 
         # Enforce the specified C++ standard
         if spec.variants["cxxstd"].value != "default":
-            cxxstd = spec.variants["cxxstd"].value
-            options.append(self.define("CMAKE_CXX_STANDARD", "{0}".format(cxxstd)))
+            options.append(self.define_from_variant("CMAKE_CXX_STANDARD", "cxxstd"))
 
         # Performance
         # Set recommended flags for maximum (matrix-free) performance, see

